### ISSUE TITLE
Problem running playbook with roles with path that includes non-ascii characters

### ISSUE THREAD

**Issue by smatyas** [ansible/ansible#10773]
I'm just learning ansible, and my problem seems a bit similar to #2752 and #3521, but in my case the path contains non-ascii charaters, so it's not a quoting problem I think.

If my ansible config dir is in a directory like this, it will fail:

```
/home/my_user/meló/ans-test
```

Everything was fine, until I started to separate things to roles subirs. The error popped up when I tried to include a role file from the main.yml like in the example here: https://github.com/ansible/ansible-examples/blob/master/lamp_simple/roles/web/tasks/main.yml

The error message I got is this:

```
Traceback (most recent call last):
  File "/usr/bin/ansible-playbook", line 323, in <module>
    sys.exit(main(sys.argv[1:]))
  File "/usr/bin/ansible-playbook", line 263, in main
    pb.run()
  File "/usr/lib/python2.7/site-packages/ansible/playbook/__init__.py", line 310, in run
    play = Play(self, play_ds, play_basedir, vault_password=self.vault_password)
  File "/usr/lib/python2.7/site-packages/ansible/playbook/play.py", line 188, in __init__
    self._tasks      = self._load_tasks(self._ds.get('tasks', []), load_vars)
  File "/usr/lib/python2.7/site-packages/ansible/playbook/play.py", line 661, in _load_tasks
    loaded = self._load_tasks(data, mv, role_params, default_vars, included_become_vars, list(included_additional_conditions), original_file=include_filename, role_name=new_role)
  File "/usr/lib/python2.7/site-packages/ansible/playbook/play.py", line 654, in _load_tasks
    include_filename = utils.path_dwim(dirname, include_file)
  File "/usr/lib/python2.7/site-packages/ansible/utils/__init__.py", line 339, in path_dwim
    return os.path.abspath(os.path.join(basedir, given))
  File "/usr/lib/python2.7/posixpath.py", line 80, in join
    path += '/' + b
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 28: ordinal not in range(128)
```

If I move everything to an only-ascii dir path, everything works fine. i.e. /home/my_user/ascii-dir/ans-test

**Comment by smatyas** [comment]
Playing with it locally it seems, that the following patch resolves my particular problem:

(under lib/ansible/utils)

```
diff __init__.py.orig __init__.py
339c339
<         return os.path.abspath(os.path.join(basedir, given))
---
>         return os.path.abspath(os.path.join(str(basedir), str(given)))
```

The idea came from here: https://code.djangoproject.com/ticket/19357
There are more `os.path.join` call in this file, and I'm not sure how to fix it in the clean way, but I hope you will :)
I'm very new to python (and ansible as well), therefore I don't want to make a pull request now.

**Comment by jimi-c** [comment]
Interesting, still appears to be an issue on the devel branch, partly because the calls to os.path in `unfrackpath()` don't like the mixed unicode/ascii:

<pre>
Traceback (most recent call last):
  File "/data/devel/ansible/bin/ansible-playbook", line 77, in <module>
    sys.exit(cli.run())
  File "/data/devel/ansible/lib/ansible/cli/playbook.py", line 162, in run
    results = pbex.run()
  File "/data/devel/ansible/lib/ansible/executor/playbook_executor.py", line 75, in run
    pb = Playbook.load(playbook_path, variable_manager=self._variable_manager, loader=self._loader)
  File "/data/devel/ansible/lib/ansible/playbook/__init__.py", line 47, in load
    pb._load_playbook_data(file_name=file_name, variable_manager=variable_manager)
  File "/data/devel/ansible/lib/ansible/playbook/__init__.py", line 78, in _load_playbook_data
    entry_obj = Play.load(entry, variable_manager=variable_manager, loader=self._loader)
  File "/data/devel/ansible/lib/ansible/playbook/play.py", line 103, in load
    return p.load_data(data, variable_manager=variable_manager, loader=loader)
  File "/data/devel/ansible/lib/ansible/playbook/base.py", line 176, in load_data
    self._attributes[name] = method(name, ds[name])
  File "/data/devel/ansible/lib/ansible/playbook/play.py", line 212, in _load_roles
    role_includes = load_list_of_roles(ds, variable_manager=self._variable_manager, loader=self._loader)
  File "/data/devel/ansible/lib/ansible/playbook/helpers.py", line 115, in load_list_of_roles
    i = RoleInclude.load(role_def, current_role_path=current_role_path, variable_manager=variable_manager, loader=loader)
  File "/data/devel/ansible/lib/ansible/playbook/role/include.py", line 50, in load
    return ri.load_data(data, variable_manager=variable_manager, loader=loader)
  File "/data/devel/ansible/lib/ansible/playbook/base.py", line 163, in load_data
    ds = self.preprocess_data(ds)
  File "/data/devel/ansible/lib/ansible/playbook/role/definition.py", line 79, in preprocess_data
    (role_name, role_path) = self._load_role_path(role_name)
  File "/data/devel/ansible/lib/ansible/playbook/role/definition.py", line 146, in _load_role_path
    role_path = unfrackpath(os.path.join(path, role_name))
  File "/usr/lib64/python2.7/posixpath.py", line 80, in join
    path += '/' + b
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 23: ordinal not in range(128)
</pre>

**Comment by jimi-c** [comment]
Actually this is in the join() call, even before it gets into the Ansible code.

**Comment by jimi-c** [comment]
# Closing This Ticket

Hi!

We believe the above commit should resolve this problem for you. This will also be included in the next major release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!

**Comment by vijayranker** [comment]
TASK: [ntpd | ntp | Configure ntp (/etc/ntp.conf)] ***************************\* 
Traceback (most recent call last):
  File "/usr/bin/ansible-playbook", line 323, in <module>
    sys.exit(main(sys.argv[1:]))
  File "/usr/bin/ansible-playbook", line 263, in main
    pb.run()
  File "/usr/lib/pymodules/python2.7/ansible/playbook/**init**.py", line 348, in run
    if not self._run_play(play):
  File "/usr/lib/pymodules/python2.7/ansible/playbook/__init__.py", line 788, in _run_play
    if not self._run_task(play, task, False):
  File "/usr/lib/pymodules/python2.7/ansible/playbook/__init__.py", line 496, in _run_task
    results = self._run_task_internal(task)
  File "/usr/lib/pymodules/python2.7/ansible/playbook/__init__.py", line 432, in _run_task_internal
    become_pass=task.become_pass,
  File "/usr/lib/pymodules/python2.7/ansible/runner/__init__.py", line 258, in **init**
    utils.plugins.push_basedir(self.basedir)
  File "/usr/lib/pymodules/python2.7/ansible/utils/plugins.py", line 33, in push_basedir
    basedir = os.path.realpath(basedir)
  File "/usr/lib/python2.7/posixpath.py", line 383, in realpath
    return abspath(path)
  File "/usr/lib/python2.7/posixpath.py", line 371, in abspath
    cwd = os.getcwd()
OSError: [Errno 2] No such file or directory

I am getting this

ansible 1.9.0.1

**Comment by jimi-c** [comment]
@vijayranker this fix will be in 2.0.
